### PR TITLE
Sink service client & unit tests

### PR DESCRIPTION
**Pull request by oschaaf**
Add the proto definition for the sink service plus a client implementation and test.

Signed-off-by: Otto van der Schaaf <ovanders@redhat.com>

### CODE REVIEW

**Comment by mum4k** [comment]
@wjuan-AFK please review and assign back to me once done.

File: source/common/nighthawk_sink_client_impl.cc
Comment (wjuan-AFK) [inline source/common/nighthawk_sink_client_impl.cc:45]: I'm not sure I understand how this loop is supposed to work.

Comment (oschaaf) [reply source/common/nighthawk_sink_client_impl.cc:45]: Ah yes, this isn't obvious; the way we handle this reflects other places in the code base, e.g:
https://github.com/envoyproxy/nighthawk/blob/3881261ccc76e382109fdf94e5035b1f7f2b71e9/source/common/nighthawk_service_client_impl.cc#L28

Some more context on that:

At the proto api level we support returning a stream of results. The sink service proto api reflects this, and accepts what NighthawkService. ExecutionStream returns as a parameter (though we wrap it in StoreExecutionRequest messages for extensibility purposes). So this implies a stream, and not a single message.

Having said that, today we constrain what we return to a single message in the implementations where this is relevant. That's why we assert here, to make sure that stays put until an explicit choice is made otherwise.

Why do this? The intent of `NighthawkService. ExecutionStream` was to be able to stream intermediate updates some day. So having streams in the api's keeps the door open on streaming intermediary updates, without forcing a change the proto api.

Comment (wjuan-AFK) [reply source/common/nighthawk_sink_client_impl.cc:45]: Thanks for the explanation.

File: api/client/service.proto
Comment (mum4k) [inline api/client/service.proto:70]: Would you mind moving the sink related service and its messages to a separate proto file for clarity? Say `sink_service.proto`?

File: api/client/service.proto
Comment (mum4k) [inline api/client/service.proto:61]: Assuming this is required, can we clearly indicate that it is?

```
// Unique id for lookup purposes. Required.
```

Alternatively if it isn't, can we describe the behavior when an empty string is provided?

File: include/nighthawk/common/nighthawk_sink_client.h
Comment (mum4k) [inline include/nighthawk/common/nighthawk_sink_client.h:26]: Why do we need to abstract the client behind another interface? Are we going to have multiple implementations?

If it is just for testing, we could do with just one implementation that has virtual methods? The additional layer of indirection comes with its cost, so trying to understand if we have a good reason to pay it.

File: include/nighthawk/common/nighthawk_sink_client.h
Comment (mum4k) [inline include/nighthawk/common/nighthawk_sink_client.h:27]: Would a reference work better? (since it cannot be null).

Looks like the other method does accept a reference.

File: source/common/nighthawk_sink_client_impl.cc
Comment (mum4k) [reply source/common/nighthawk_sink_client_impl.cc:45]: Can we convert a portion of this thread into an inline comment?

Comment (mum4k) [inline source/common/nighthawk_sink_client_impl.cc:45]: Why did we choose an ASSERT as compared to returning not-ok status? Is crashing the best choice here?

File: test/common/nighthawk_sink_client_test.cc
Comment (mum4k) [inline test/common/nighthawk_sink_client_test.cc:160]: Did we mean to leave this in?

**Review by mum4k — CHANGES_REQUESTED** [review]

File: api/client/service.proto
Comment (oschaaf) [reply api/client/service.proto]: Done in fe2eac357d600922489e611a8843548b544a16dd. Let me know what you think, happy to iterate more on this.

File: api/client/service.proto
Comment (oschaaf) [reply api/client/service.proto]: Done: d89620e32d3b28caf5fd1eab11b8bff794bdcbca

File: include/nighthawk/common/nighthawk_sink_client.h
Comment (oschaaf) [reply include/nighthawk/common/nighthawk_sink_client.h]: Good catch; that was pretty inconsistent. Done in fe2eac357d600922489e611a8843548b544a16dd

Comment (oschaaf) [reply include/nighthawk/common/nighthawk_sink_client.h]: Err sorry f64a229b62bcaaa78b1316708fca4cf4324c3a1d

File: source/common/nighthawk_sink_client_impl.cc
Comment (oschaaf) [reply source/common/nighthawk_sink_client_impl.cc:59]: Done in 3734e54217e9a615c117f73b69f213902cd82f25

File: test/common/nighthawk_sink_client_test.cc
Comment (oschaaf) [reply test/common/nighthawk_sink_client_test.cc]: Remove in 78dad245c0fd4e3f907296efba461ecaaf9b7e7d

File: source/common/nighthawk_sink_client_impl.cc
Comment (oschaaf) [reply source/common/nighthawk_sink_client_impl.cc:59]: This is something that really should not ever happen unless we explicitly decided otherwise. Having said that, personally I would lean slightly towards returning not ok and then cover this in a test as a behavioural aspect; but for now I settled for this because I think consistency is outweighing this (as this mirrors  another part of the code base).

File: include/nighthawk/common/nighthawk_sink_client.h
Comment (oschaaf) [reply include/nighthawk/common/nighthawk_sink_client.h]: This makes it easy to create a mock for it; so if `ProcessImpl` starts using this further down the road, this comes in handy when we want to unit test the interaction to ensure that looks like we expect it to. But I think you are right in that chances are pretty slim that we'll see >1  implementations for this :-)
 
If you have a concern about perf aspects, imho that only weighs slightly here since this will not be something that will be repeatedly consumed at a frequency where this would become significant; but I can see that there's potential overhead in maintenance of the interface layer.. But here there's also a consistency argument: other similar code does this too.

Weighing everything, I'm personally slightly leaning towards leaving this as is; but I'll ping you on Slack to discuss some more, as changing this is also not something that I'd strongly disagree with.

File: api/sink/sink.proto
Comment (mum4k) [inline api/sink/sink.proto:3]: In our last discussion about namespaces, we ended up on having just a single namespace for Nighthawk's future code. Do we have a good reason to introduce the new namespace here?

File: include/nighthawk/common/nighthawk_sink_client.h
Comment (mum4k) [reply include/nighthawk/common/nighthawk_sink_client.h]: A class with virtual methods should make it as easy to create mocks. The interface tends to only be really needed when there are multiple concrete implementations. That is also how a reader of the code "reads" the existence of a virtual interface. Having it just for mocking may be considered too big of a hammer.

With that said, we don't need to block this PR on it, it is something we can start considering and applying on future code. Let's proceed as is for expediency.

File: source/common/nighthawk_sink_client_impl.cc
Comment (mum4k) [reply source/common/nighthawk_sink_client_impl.cc:59]: Happy to leave this up to you. For the sake of future discussions - this may not be a strong case for consistency. Consistency across the codebase is very important on items like APIs, used libraries, namespaces, etc. Individual functions should be able to make their own choice of whether they crash the code or return an error. I have learned to lean towards error returning, since crashes are harder to document and are often surprising to someone who might end up reusing the library or the function. Additionally crashes are not the best experience for our users.

With that said, I agree with you in this case that since we are making a choice to only return a single message, returning multiple is technically a programming error which does deserve a crash.

**Review by mum4k — APPROVED** [review]

**Comment by mum4k** [comment]
@oschaaf just one remaining question about the new namespace.

File: api/sink/sink.proto
Comment (oschaaf) [reply api/sink/sink.proto]: Modified in 5c49c607558f7fb26fb088c787b8cb22dd820e5d

**Review by mum4k — APPROVED** [review]

**Pull request merged**
